### ISSUE TITLE
Race in StreamObservers.copyWithFlowControl

### ISSUE THREAD

**Issue by DRayX** [grpc/grpc-java#6786]
### What version of gRPC-Java are you using?
1.28

### What is your environment?

Linux Java 8

### What did you expect to see?
That StreamObservers.copyWithFlowControl would deterministically copy values

### What did you see instead?
That StreamObservers.copyWithFlowControl has a race that can lead to non-deterministic behavior

### Steps to reproduce the bug
The following test class reproduces the bug (generally after a hundred thousand attempts or so)

```java
public final class StreamObserversRace {
  private static final ImmutableList<Integer> LIST =
      ImmutableList.of(0, 1, 2, 3, 4, 5, 6, 7, 8, 9);

  static final class TestStreamObserver<T> extends CallStreamObserver<T> {
    final ConcurrentLinkedQueue<T> values = new ConcurrentLinkedQueue<>();
    final ConcurrentLinkedQueue<FutureTask<Void>> waits = new ConcurrentLinkedQueue<>();
    final FutureTask<Void> completed = new FutureTask<>(() -> {}, null);
    private final Executor executor;
    private Runnable onReady;
    private volatile boolean ready = true;

    TestStreamObserver(Executor executor) {
      this.executor = executor;
    }
    @Override public boolean isReady() {
      return ready;
    }
    @Override public void setOnReadyHandler(Runnable onReady) {
      this.onReady = onReady;
      executor.execute(onReady);
    }
    @Override public void onNext(T value) {
      this.ready = false;
      FutureTask<Void> wait = new FutureTask<>(() -> {}, null);
      waits.add(wait);
      executor.execute(() -> {
        values.add(value);
        this.ready = true;
        onReady.run();
        wait.run();
      });
    }
    @Override public void onCompleted() {
      executor.execute(completed);
    }
    @Override public void onError(Throwable t) {}
    @Override public void disableAutoInboundFlowControl() {}
    @Override public void request(int count) {}
    @Override public void setMessageCompression(boolean enable) {}
  }

  static void main(String[] args) throws Exception {
    ExecutorService executor = Executors.newCachedThreadPool();
    for (int i = 0;; ++i) {
      TestStreamObserver<Integer> test = new TestStreamObserver<>(executor);
      StreamObservers.copyWithFlowControl(LIST, test);
      test.completed.get();
      for (FutureTask<Void> wait : test.waits) {
        wait.get();
      }
      if (!LIST.equals(ImmutableList.copyOf(test.values))) {
        System.out.println(String.format("Failed after %d attempts (%s)", i, test.values));
        break;
      }
    }
    executor.shutdown();
  }
}
```

### Why this race happens
If the `FlowControllingOnReadyHandler.run` method is interrupted between `target.onNext` and the next `target.isReady` call, it is possible for it to never see that the target was in a not ready state, meaning that both a new invocation of the ready handler will be started, and the current one will continue.  In this test this is exhibited by the list being out of order.

### General issues with the ready mechanism
This race is caused by a more general design flaw of the `isReady`/`onReady` mechanism used across `ClientCall`, `ServerCall`, and `CallStreamObserver`.  The typical pattern of 

```java
@Override public void onReady() {
  while(observer.isReady()) {
    observer.onNext(getMessage());  // or call.sendMessage
    // Issue happens if the thread is interrupted here and the
    // stream transitions from not-ready to ready.
  }
}
```

always has this race of the stream transitioning from not ready to ready between the call to `next` and the call to `isReady`.  Avoiding this race is very non-trivial, especially given the limited invariants of the onReady call (there doesn't appear to be any guarantee that this method isn't called spuriously when the stream was already ready).  I've been trying to find a good example of this for a while, but this is the first concrete example I've come up with.

### Potential solution
One potential fix would be to have `StreamObserver.onNext` (and `{Client|Server}Call.sendMessage`) return a boolean atomically indicating that the underlying stream is ready, and to add an invariant that `onReady` will be called if and only if the underlying stream transitions from a state where it is visibly not ready to one where it is, but this would be a breaking change on any implementations of these classes.  The above pattern would then become

```java
@Override public void onReady() {
  while (observer.onNext(getMessage())) {}
}
```

Note that this solution only works when the messages are being pulled.  When the messages are also being pushed asynchronously, rendezvousing the message and the ready event would still be tricky.  For this use case, a `whenReady(Runnable)` mechanism that invokes the handler exactly once (immediately if already ready) would be preferable.  This approach while more general would be more expensive than the one above as it requires creating a runnable for each message that is sent.

**Comment by creamsoup** [comment]
the problem here is the `onReadyHandler` is running on the current thread instead of the executor. This makes 2 threads fighting each other. 

if you change the `setOnReadyHandler` to use executor,  the race should be fixed.

```java
    @Override public void setOnReadyHandler(Runnable onReady) {
      this.onReady = onReady;
      executor.execute(onReady);
    }
```

**Comment by DRayX** [comment]
That wasn't really my point; the fact of the matter is that bytes being flushed to the channel can happen on a different thread.  A better example would use `Executors.newCachedThreadPool` (the default used for `SHARED_CHANNEL_EXECUTOR` used by default in both servers and channels), I was just lazy and used the main thread as the second thread.  I've updated the above test (which also avoids the false positive of out of order completion).

**Comment by creamsoup** [comment]
the `StreamObserver` returned from grpc should be running in serial order. because the call wraps executor in `SerializingExecutor` (or directExecutor which is serial anyways). also, all the channel operations are under netty's event loop, only 1 thread is involved for given channel. So, i think your repro above is not really reflecting actual environment. 

I wouldn't be surprised if there's such a rare race, are you seeing this issue in your application?

**Comment by DRayX** [comment]
Oh, OK, that actually makes sense, and the documentation on setOnReadyHandler does actually cover this, so you're right, there likely isn't a race in this case (fully blocking use of onReady).  Should probably go ahead an close this issue, as I don't think there is a race in this method.

If the messages are generated asynchronously, it seems much more likely that one would run into a similar issue.  Imagine using a Reactive Streams `Publisher` instead of an `Iterable`:

```java
public static <V> void copyWithFlowControl(Publisher<V> source, CallStreamObserver<V> target) {
  source.subscribe(new Subscriber<V>() {
    Subscription subscription;
    @Override public void onSubscribe(subscription s) {
      this.subscription = s;
      target.setOnReadyHandler(this::checkReady);
    }
    @Override public void onNext(V message) {
      target.onNext(message);
      checkReady();
    }
    @Override public void onComplete() {
      target.onCompleted();
    }
    @Override public void onError(Throwable error) {
      target.onError(error);
    }
    void checkReady() {
      if (target.isReady()) {
        subscription.request(1);
      }
    }
  });
}
```

This may actually more or less work out in that all of the messages should be delivered in order due to the serialization guarantees of reactive streams, but will lead to excessive buffering as multiple messages may be requested from the `Publisher` for the stream being ready once.  If the interface didn't support concurrent requests (e.g. Iterable<ListenableFuture<T>>) this would pretty obviously fall over.

My general point is that ensuring a single "readiness" (either from isReady or an onReady handler) is very complicated with the API as is (hence why a `whenReady` method could alternatively solve this issue, but with more overhead than an atomic ready state from onNext).  Deep in the internals, `AbstractStream.TransportState.onSendingBytes` could check readiness with the onReadyLock held adding almost no overhead, but passing this up through the stack is a decent amount of work (and it's super unclear how RetriableStream would handle this).

I ran into a similar issue in production once in a much more complicated use case where I actually missed the ready event, but I worked around that with a couple of atomics, and I can't think of an easy repro.

Additionally, `{Client|Server}Call.Listener.onReady` don't seem to have the same guarantee (although I imagine they do, it just isn't documented).
